### ISSUE TITLE
Crash "CFRelease() called with NULL" on MacOS with bold fonts like "Arial"

### ISSUE THREAD

**Issue by jmw86069** [r-lib/marquee#44]
My fault for working in RStudio using the ragg device!
I've been digging to try to find the cause... best I can do is that the Apple bug report includes this error:
`"CFRelease() called with NULL"`

It seems that using `ragg` for output is solid (and looks great!), both in RStudio and saving to file.
Other graphics devices, including Quartz, `png()`, `cairo_pdf()` immediately crash R.

### Test case that does not crash, using "sans" font:
```R
use_style_set1 <- marquee::classic_style(body_font="sans", base_size=18);
mg1 <- marquee::marquee_grob(
   text="Set **Label** Here",
   x=0.5, y=0.5,
   style=use_style_set1);
grid::grid.newpage();
grid::grid.draw(mg1);
```
<img width="258" alt="image" src="https://github.com/user-attachments/assets/4c8fa4dc-56d9-4bb4-9797-ed9303d48efb">

(Notice it doesn't show the bold text with default Quartz output on MacOS. If using RStudio with ragg output, it shows the bold text.Might be a clue, idk.)

### Test case that crashes, using "Arial" font:
```R
use_style_set2 <- marquee::classic_style(body_font="Arial");
mg2 <- marquee::marquee_grob(
   text="Set **Label** Here",
   x=0.5, y=0.5,
   style=use_style_set2);
grid::grid.newpage();
grid::grid.draw(mg2);
```

The R console output looks like this:
```
> grid::grid.draw(mg2);
Trace/BPT trap: 5
```

### Other comments
It also crashes when trying to save with `png()`:
```R
# crashes
png("test_png.png")
grid::grid.newpage();
grid::grid.draw(mg2);
dev.off()
```

However **no crash** when saving with `ragg::agg_png()`:
```R
# no crash
ragg::agg_png("test_agg_png.png")
grid::grid.newpage();
grid::grid.draw(mg2);
dev.off()
```
(Fwiw, ragg output shows proper bold text.)
<img width="157" alt="image" src="https://github.com/user-attachments/assets/b08ab8ec-d99f-427b-9611-4602d0476dc6">

I suspect the `ragg` device is more tolerant of whatever is happening, other devices not so much.
Since you also wrote ragg, you may uniquely recognize what feature you added to ragg? :)

I have fresh install of R-4.3.3. I did use `rig` to install this version of R, so the installation is fairly clean. However, I have no idea if `rig` would interfere somehow with dylib library files. I can try R-4.4.2 if that might help.

I don't get these errors using `grid::textGrob()`, nor with `gridtext::richtext_grob()`. I think it's something specific to `marquee`, or its dependencies: `systemfonts`, `textshaping`.

It crashes for TTF and TTC fonts, not specific to either.
It also crashes if I use `weight="bold"` and without markdown text.

I tried running `textshaping::shape_text(c("Test ", "Text", " Here"), family="Arial", weight=c("normal", "bold", "normal"))` and it returned all metrics I'd expect, each font was correctly identified ("Arial.ttf" and "Arial Bold.ttf") for example.

Thanks for your time, when you have time to take a look that is. Due respect!

**Comment by jmw86069** [comment]
TL;DR - sorry for the noise. *sigh*
The error goes away when using R-4.4.2. So... marquee may effectively require R-4.4.2 and higher on MacOS. (Without a fix, see below for small suggestion.) :)

For R-4.3.3, I did some sleuthing:
RQuartz_glyph->CFRelease
<img width="615" alt="image" src="https://github.com/user-attachments/assets/a7e59ba5-3ec5-44d0-bce8-0a8d1001549b">

Seems related to this fix in R-4.4.2: fixes crash when the font file path has whitespace.
(I didn't notice till just now that the crash happens only when the font file path has whitespace. Geez, I'm so bad at that.)
https://github.com/wch/r-source/commit/6c3bf4a5af82ee75ca7e3e22f5b401e051e1e835

I answered my own next question (thank goodness right?)
Yes the only way to get the `grid` fix is to update to R-4.4.2.

So...
My small suggestion is for marquee (or systemfonts) to avoid crashing R on MacOS R-4.4.1 and older:
* if check R.version < 4.4.2, `compareVersion("4.4.2", paste(R.version$major, R.version$minor, sep=".")) == 1`
* and if MacOS specific os: `(grepl("^darwin", R.version$os))` 
* and if whitespace in `systemfonts::system_fonts()` "path" column `grepl(" ", path)`
* then **don't ever use those fonts**. On MacOS R-4.4.1 and older, it will absolutely crash the R session. You could `subset(systemfonts::system_fonts(), !grepl(" ", path))` then use that?

If I understand right, hopefully it will pick an alternate font? Better than R crash.

**Comment by jmw86069** [comment]
Thank you for the update!

TL;DR: For the original test case, MacOS running R-4.3.3, it still crashes the R session.
Binary packages are not available for marquee-1.0.0 on R-4.3.3, nor its dependency systemfonts-1.2.0.
I also could not install systemfonts-1.2.0 from source.
So the workaround is not available for R-4.3.3 where it is needed.

If marquee will only work on R-4.4.0 that's unfortunate. (I'd really love to use it!)

## The problem I'm trying to solve by using marquee.

I want to draw grid text `'\u2191\u2193 215'` which should look like this:
'&#x2191;&#x2193; 215'
Fonts don't reliably unicode glyphs like upArrow/downArrow, so in a variety of scenarios they render as empty boxes:
'&#x2610;&#x2610; 215'

On each system (MacOS, linux, Windows), even with fonts that should have Unicode support, it only works with certain fonts (displaying proper Unicode arrows). For me, it needs "Arial" on MacOS, "Helvetica" on linux, and idk on Windows.
And it varies by LOCALE, and by device. The 'ragg' device appears to "work" without crashing, it just works, and it's awesome. But I also need PDFs.
Any font with a space in the path (e.g. "Arial Bold") on MacOS causes R to crash when using marquee.

Is there any way to determine if a unicode character `'\u2191'` will properly resolve to an upArrow beforehand? Programmatically I could change to `^` and `v` as a cheap hack. Anything is better than &#x2610;.

## Installing marquee from source

Trying to install from source fails due to r-lib/systemfonts#122, the workaround is to update OBJCXXFLAGS.

Can that step be automated? I think it would fix the CRAN build issues here: https://www.r-project.org/nosvn/R.check/r-oldrel-macos-arm64/systemfonts-00install.html 

So I tried installing from source with the OBJCXXFLAGS:
```R
withr::with_makevars(c(OBJCXXFLAGS = "${CXX17STD}"),
   install.packages('marquee', type = 'source'))
```

It seemed to compile systemfonts properly, then failed to load the `systemfonts.so` file.
It seems to ignore the path to libfreetype.6.dylib, maybe `@rpath` is incorrect here? Total guess.

<details>
        <summary>Error installing systemfonts from source (Click to Expand)</summary>

```
* installing *source* package ‘systemfonts’ ...
** package ‘systemfonts’ successfully unpacked and MD5 sums checked
** using staged installation
Found pkg-config cflags and libs!
Using PKG_CFLAGS=-I/Users/wardjm/mambaforge/include/freetype2
Using PKG_LIBS=-L/Users/wardjm/mambaforge/lib -lfreetype -lz -lpng16 -lz
** libs
using C++ compiler: ‘Apple clang version 16.0.0 (clang-1600.0.26.4)’
using SDK: ‘MacOSX15.1.sdk’
rm -f systemfonts.so caches.o cpp11.o dev_metrics.o font_matching.o font_local.o font_registry.o ft_cache.o string_shape.o font_metrics.o font_outlines.o font_fallback.o string_metrics.o emoji.o cache_store.o init.o mac/FontManagerMac.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c caches.cpp -o caches.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c cpp11.cpp -o cpp11.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c dev_metrics.cpp -o dev_metrics.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c font_matching.cpp -o font_matching.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c font_local.cpp -o font_local.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c font_registry.cpp -o font_registry.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c ft_cache.cpp -o ft_cache.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c string_shape.cpp -o string_shape.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c font_metrics.cpp -o font_metrics.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c font_outlines.cpp -o font_outlines.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c font_fallback.cpp -o font_fallback.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c string_metrics.cpp -o string_metrics.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c emoji.cpp -o emoji.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c cache_store.cpp -o cache_store.o
clang++ -arch arm64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c init.cpp -o init.o
clang++ -arch arm64 -I"/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/include" -DNDEBUG -I/Users/wardjm/mambaforge/include/freetype2 -I'/Users/wardjm/Library/R/arm64/4.3/library/cpp11/include' -I/opt/R/arm64/include   -fobjc-arc -fPIC  -std=gnu++17 -c mac/FontManagerMac.mm -o mac/FontManagerMac.o
clang++ -arch arm64 -std=gnu++17 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -L/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib -L/opt/R/arm64/lib -o systemfonts.so caches.o cpp11.o dev_metrics.o font_matching.o font_local.o font_registry.o ft_cache.o string_shape.o font_metrics.o font_outlines.o font_fallback.o string_metrics.o emoji.o cache_store.o init.o mac/FontManagerMac.o -L/Users/wardjm/mambaforge/lib -lfreetype -lz -lpng16 -lz -framework CoreText -framework Foundation -F/Library/Frameworks/R.framework/Versions/4.3-arm64 -framework R -Wl,-framework -Wl,CoreFoundation
ld: warning: ignoring duplicate libraries: '-lz'
installing to /Users/wardjm/Library/R/arm64/4.3/library/00LOCK-systemfonts/00new/systemfonts/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for ‘systemfonts’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/Users/wardjm/Library/R/arm64/4.3/library/00LOCK-systemfonts/00new/systemfonts/libs/systemfonts.so':
  dlopen(/Users/wardjm/Library/R/arm64/4.3/library/00LOCK-systemfonts/00new/systemfonts/libs/systemfonts.so, 0x0006): Library not loaded: @rpath/libfreetype.6.dylib
  Referenced from: <E64A0173-DBB9-330C-82C6-46C9DE3D36C9> /Users/wardjm/Library/R/arm64/4.3/library/00LOCK-systemfonts/00new/systemfonts/libs/systemfonts.so
  Reason: tried: '/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libfreetype.6.dylib' (no such file), '/Library/Java/JavaVirtualMachines/jdk-11.0.18+10/Contents/Home/lib/server/libfreetype.6.dylib' (no such file)
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/Users/wardjm/Library/R/arm64/4.3/library/systemfonts’
* restoring previous ‘/Users/wardjm/Library/R/arm64/4.3/library/systemfonts’
```

</details>

A snippet of the error is shown below:
```
Library not loaded: @rpath/libfreetype.6.dylib
  Referenced from: <E64A0173-DBB9-330C-82C6-46C9DE3D36C9> /Users/wardjm/Library/R/arm64/4.3/library/00LOCK-systemfonts/00new/systemfonts/libs/systemfonts.so
  Reason: tried: '/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libfreetype.6.dylib' (no such file), '/Library/Java/JavaVirtualMachines/jdk-11.0.18+10/Contents/Home/lib/server/libfreetype.6.dylib' (no such file)
```

Thanks for your help or suggestions!
